plugins/modules/ec2_vpc_endpoint_service.py

```diff
@@ -0,0 +1,513 @@
+"""
+
+try:
+    import botocore
+except ImportError:
+    pass  # Handled by AnsibleAWSModule
+
+import json
+from datetime import datetime
+from typing import Optional
+
+from ansible_collections.amazon.aws.plugins.module_utils.core import AnsibleAWSModule
+from ansible_collections.amazon.aws.plugins.module_utils.core import is_boto3_error_code
+from ansible_collections.amazon.aws.plugins.module_utils.ec2 import AWSRetry
+from ansible_collections.amazon.aws.plugins.module_utils.ec2 import ansible_dict_to_boto3_filter_list
+from ansible_collections.amazon.aws.plugins.module_utils.ec2 import boto3_tag_list_to_ansible_dict
+from ansible.module_utils.common.dict_transformations import camel_dict_to_snake_dict
+from ansible.module_utils.common.dict_transformations import snake_dict_to_camel_dict
+
+from ansible.module_utils.resource_tags import ensure_tags
+from ansible_collections.amazon.aws.plugins.module_utils.tagging import ansible_dict_to_boto3_tag_list  # pylint: disable=unused-import
+
+try:
+    from botocore.exceptions import ClientError, BotoCoreError
+except ImportError:
+    pass  # Handled by AnsibleAWSModule
+
+ARGUMENT_SPEC = dict(
+    state=dict(type="str", choices=['present', 'absent'], default='present'),
+    private_dns_name=dict(required=False, type='str'),
+    supported_ip_address_types=dict(required=False, type='list', default=['ipv4']),
+    gateway_load_balancer_arns=dict(required=False, type='list', default=[]),
+    network_load_balancer_arns=dict(required=False, type='list', default=[]),
+    permissions=dict(required=False, type='list', default=[]),
+    purge_permissions=dict(default=False, type='bool'),
+    tags=dict(required=False, type='dict', aliases=['resource_tags']),
+    acceptance_required=dict(default=False, type='bool'),
+    service_id=dict(required=False, type='str'),
+)
+
+REQUIRED_IF = [
+    ('state', 'absent', ['service_id']),
+]
+
+SUPPORTS_CHECK_MODE = True
+
+
+def format_client_params(
+    module: AnsibleAWSModule,
+    endpoint_service: dict,
+    tags: Optional[dict] = None,
+    service_id: Optional[str] = None,
+    operation: Optional[str] = None,
+) -> dict:
+    """
+    Formats plan details to match boto3 backup client param expectations.
+
+    module : AnsibleAWSModule object
+    endpoint_service: Dict of endpoint service details
+    tags: Dict of endpoint service tags
+    service_id: ID of endpoint service to update, only needed for update operation
+    operation: Operation to add specific params for, either create or update
+    """
+
+    params = snake_dict_to_camel_dict(
+        {k: v for k, v in endpoint_service.items() if k not in ['permissions', 'purge_permissions', 'service_id', 'update_needed'] and v is not None},
+        capitalize_first=True,
+    )
+
+    if operation == "create":  # Add create-specific params
+        if tags:
+            params["TagSpecifications"] = [{'ResourceType': 'vpc-endpoint-service', 'Tags': ansible_dict_to_boto3_tag_list(tags)}]
```

Comment:
`ansible_collections.amazon.aws.plugins.module_utils.tagging.boto3_tag_specifications` can generate these for you.

Reply:
this part of code I took again from backup_plan, so first I need to understand what the "poet means"  and adopt

Reply:
The AWS APIs have a rather weird way of specifying Tags (and they're not even consistent)

Let's say you wanted to apply the following tags (in usual Key/Value format
```
{
  "MyFirstKey": "My first value", 
  "my_second_key": "another value",
)
```

There are two ways the EC2 API want you to pass this 

1. A "list' of tags" aka `boto3_tag_list` or list-of-dicts (mostly when updating tags):
  ```
  [
    {
      "Key": "MyFirstKey",
      "Value": "My first value",
    },
    {
      "Key": "my_second_key",
      "Value": "another value",
    }
  ]
  ```
2. `TagSpecifications` - a mapping of resource types to a boto3_tag_list (mostly when creating 1 or more resources)
  ```
  [
    {
      "ResourceType": "vpc-endpoint-service",
      "Tags": [
        {
          "Key": "MyFirstKey",
          "Value": "My first value",
        },
        {
          "Key": "my_second_key",
          "Value": "another value",
        }
      ]
    }  
  ]
  ```

These can be something of a PITA to work with, and Amazon's not even consistent with the formats it uses :(

However, we have some helpers to try and make this easier to deal with.
* ansible_dict_to_boto3_tag_list - Takes a simple dictionary and converts it to the `boto3_tag_list` format
* boto3_tag_list_to_ansible_dict - The other way around
* boto3_tag_specifications - Takes a list of resource types and a simply dictionary representing the tags, then returns the `TagSpecifications` format.

What gets really messy is that not all of Amazon's APIs even use the same formats: some use simple dicts, some the above format, and some even use list-of-dicts change "Key" and "Value"